### ISSUE TITLE
no-fallthrough should understand statement reachability

### ISSUE THREAD

**Issue by peterhal** [eslint/eslint#4246]
The following currently generates no-fallthrough warnings and should not:

```js
switch (c) {
  case 'a': // should not generate no-fallthrough as the end of the if statement is not reachable
     if (c2) { return; } else {return; }
  case 'b';
    break;
}
```

```js
switch (c) {
  case 'a': // should no generate no-fallthrough, but does.
    switch (c2) {
      case 'a':
        return;
      default:
        return;
      }
  case 'b':
    break;
}
```

The following statements need to be handled to correctly understand control flow:

`return`, `throw`, `break`, `continue`, `for`, `while`, `do`, `if`, `try`, `=>`, `function`, `switch`

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
This is a known issue. This will be addressed once #3559 is going to be merged.

**Comment by adambuczynski** [comment]
@ilyavolodin this seems to have been broken again in a recent version of eslint.
I'm getting a `no-fall-through` error for code with a simple return:

```js
case 'years':
        return eventDate.clone().add(frequency, interval);
```

ESLint version is 2.11.1 (bundled with Atom linter-eslint package).

**Comment by adambuczynski** [comment]
@ilyavolodin false alarm, I misinterpreted what line was causing the error, it was actually a fall through above that line.
